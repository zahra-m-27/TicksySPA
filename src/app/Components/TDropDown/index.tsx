import {useEffect, useRef, useState} from 'react';
import Assets from '../../Assets';
import styles from './styles.module.scss';
import ClassNames from '../../Utilities/ClassNames';

interface Item<T> {
  value: T;
  label: string;
}

interface Props<T> {
  label?: string;
  items: Item<T>[];
  className?: string;
  selectedItem?: T[];
  multiSelect?: boolean;
  labelClassName?: string;
  onSelect?: (values: T[]) => void;
}

export default function TDropDown<T>({
  label,
  items,
  onSelect,
  className,
  multiSelect,
  selectedItem,
  labelClassName,
}: Props<T>) {
  const [Open, setOpen] = useState<boolean>(false);
  const container = useRef<HTMLDivElement>(null);
  const input = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const onClickContainer = () => setOpen(false);
    const onClickInput = (e: any) => {
      e.stopPropagation();
      setOpen(!Open);
    };
    container.current?.parentElement?.addEventListener(
      'click',
      onClickContainer,
      false
    );
    input.current?.addEventListener('click', onClickInput, false);
    return () => {
      container.current?.parentElement?.removeEventListener(
        'click',
        onClickContainer
      );
      input.current?.removeEventListener('click', onClickInput);
    };
  }, [Open]);

  const label_style = ClassNames(styles.label, labelClassName);
  const container_style = ClassNames(styles.container, className);

  const selectedLabel = items
    ?.filter((i) => selectedItem?.includes(i.value))
    .map((i) => i.label)
    ?.join(', ');

  return (
    <div ref={container} data-open={Open} className={container_style}>
      <p className={label_style}>{label}</p>
      <div
        ref={input}
        className={styles.content}
        data-testid="dropdown-container">
        <img
          alt="arrow down"
          src={Assets.SVGs.ArrowDown}
          className={styles.arrow_down}
        />
        <p className={styles.selected_item_label}>
          {selectedLabel.length ? selectedLabel : 'انتخاب كنيد'}
        </p>
      </div>

      <div className={styles.items}>
        {items.map((item) => (
          <div
            key={`${item.value}`}
            className={styles.item}
            onClick={() => {
              setOpen(!!multiSelect);
              if (selectedItem) {
                const index = selectedItem.indexOf(item.value);
                if (index >= 0) {
                  onSelect &&
                    onSelect(selectedItem.filter((v, i) => i != index));
                } else {
                  onSelect && onSelect([item.value, ...selectedItem]);
                }
              } else {
                onSelect && onSelect([item.value]);
              }
            }}>
            {multiSelect &&
              selectedItem &&
              selectedItem.indexOf(item.value) >= 0 && (
                <Assets.SVGs.CheckmarkSVG className={styles.checkmark} />
              )}
            <p>{item.label}</p>
          </div>
        ))}
      </div>
    </div>
  );
}
