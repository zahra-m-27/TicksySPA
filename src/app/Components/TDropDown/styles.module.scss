.container {
  width: 100%;
  cursor: pointer;
  text-align: end;
  margin-top: 10px;
  position: relative;

  .label {
    color: #333333;
    margin-bottom: 10px;
  }

  .content {
    display: flex;
    padding: 10px 15px;
    border-radius: 5px;
    background-color: #dadde0;
    border: 1px solid #e3e3e3;

    p {
      flex: 1;
      margin: 0;
      text-align: end;
    }
  }

  .arrow_down {
  }

  &[data-open='false'] {
    .items {
      display: none;
    }
  }

  .selected_item_label {
    color: #606f7b;
    cursor: pointer;
  }

  .items {
    width: 100%;
    padding: 10px;
    cursor: pointer;
    transition: 0.6s;
    position: absolute;
    border-radius: 8px;
    background-color: #eeeff1;
    box-shadow: 0 3px 6px 0 rgb(0 0 0 / 16%);

    .item {
      color: #000;
      margin: 0.5px;
      display: block;
      text-align: end;
      padding: 8px 20px;

      &:hover {
        overflow: hidden;
        position: relative;
        border-radius: 5px;
        background-color: #dadde0;
      }
    }
  }
}
