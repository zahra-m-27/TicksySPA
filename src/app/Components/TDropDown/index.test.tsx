import TDropDown from './index';
import React from 'react';
import {render} from '@testing-library/react';

const sample = (
  <TDropDown
    items={[
      {
        value: 'val1',
        label: 'تست1',
      },
      {
        value: 'val2',
        label: 'تست2',
      },
      {
        value: 'val3',
        label: 'تست3',
      },
    ]}
  />
);
test('TDrop', () => {
  const {getByText} = render(sample);
  getByText(/تست1/i);
  getByText(/تست2/i);
  getByText(/تست3/i);
});
