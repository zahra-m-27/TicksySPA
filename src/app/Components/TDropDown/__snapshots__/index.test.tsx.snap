// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  className="container"
  data-open={false}
>
  <p
    className="label"
  >
    Test Label
  </p>
  <div
    className="content"
    data-testid="dropdown-container"
  >
    <img
      alt="arrow down"
      className="arrow_down"
      src="arrow_down.svg"
    />
    <p
      className="selected_item_label"
    >
      انتخاب كنيد
    </p>
  </div>
  <div
    className="items"
  >
    <div
      className="item"
      onClick={[Function]}
    >
      <p>
        تست1
      </p>
    </div>
    <div
      className="item"
      onClick={[Function]}
    >
      <p>
        تست2
      </p>
    </div>
    <div
      className="item"
      onClick={[Function]}
    >
      <p>
        تست3
      </p>
    </div>
  </div>
</div>
`;
